linera-sdk/src/lib.rs

```diff
@@ -67,39 +71,52 @@ pub use linera_base::ensure;
 #[doc(hidden)]
 pub use wit_bindgen_guest_rust;
 
-/// A simple state management runtime using a single byte array.
+/// A simple state management runtime based on a single byte array.
 pub struct SimpleStateStorage<A>(std::marker::PhantomData<A>);
 
-/// A state management runtime based on `linera-views`.
+/// A state management runtime based on [`linera_views`].
 pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
 
-/// The public entry points provided by an application's contract.
+/// The contract interfaces of a Linera application.
+///
+/// As opposed to the [`Service`] interfaces of an application, contract entry points
+/// are triggered by the execution of blocks in a chain. Their execution may modify
+/// storage and is gas-metered.
 ///
-/// Execution of these endpoints consume fuel, because they can change the application's state and
-/// are therefore consensus criticial.
+/// Below we use the word "transaction" to refer to the current operation or effect being
+/// executed.
 #[async_trait]
-pub trait Contract: Sized {
-    /// Message reports for application execution errors.
-    type Error: Error;
-    /// The desired storage backend to use to store the application's state.
-    type Storage;
+pub trait Contract: ContractAbi + Sized {
+    /// The type used to report errors to the execution environment.
+    ///
+    /// Errors are not recoverable and always interrupt the current transaction. To return
+    /// recoverable errors in the case of application calls and session calls, you may use
+    /// the response types.
+    type Error: Error + From<serde_json::Error> + From<bcs::Error>;
```

Comment:
I was thinking, since we now have a known set of errors that need to be handled by the user, (serde and bcs), how about we make the contract error type an enum, a bit like this:

```rust
trait Contract {
    type ExecutionError: Error;
    
    fn some_func() -> Result<(), ContractError<Self::ExecutionError>>;
}

enum ContractError<ExecutionError> {
    Serialization,
    Execution(ExecutionError)
}
```
then users won't need the boilerplate from bcs & json.

Reply:
Or maybe they should not be visible to the contract author at all:
If someone calls the contract with something that can't be deserialized, or if the contract returns something that can't be serialized, that's just a failed operation/effect, as if it had panicked?
Because I don't think the former is a case the contract should have to handle in any way, and the latter is a bug in the contract itself.

Reply:
Yes, the error management needs to revisited as suggested by #743